# PyCity Schools Analysis

- Given below is an analysis of the student grades in Reading and Math.
- This data used in this analysis corresponds to the grades of all the students that attend over 15 different schools.
  
---

In [536]:
# Dependencies and Setup
import pandas as pd
import numpy as np
from pathlib import Path

# File to Load (Remember to Change These)
school_data_to_load = Path("Resources/schools_complete.csv")
student_data_to_load = Path("Resources/students_complete.csv")

# Read School and Student Data File and store into Pandas DataFrames
school_data = pd.read_csv(school_data_to_load)
student_data = pd.read_csv(student_data_to_load)

# Combine the data into a single dataset.  
school_data_complete = pd.merge(student_data, school_data, how="left", on=["school_name", "school_name"])
school_data_complete.head()

Student ID       student_name gender grade        school_name  \
0           0       Paul Bradley      M   9th  Huang High School   
1           1       Victor Smith      M  12th  Huang High School   
2           2    Kevin Rodriguez      M  12th  Huang High School   
3           3  Dr. Richard Scott      M  12th  Huang High School   
4           4         Bonnie Ray      F   9th  Huang High School   

   reading_score  math_score  School ID      type  size   budget  
0             66          79          0  District  2917  1910635  
1             94          61          0  District  2917  1910635  
2             90          60          0  District  2917  1910635  
3             67          58          0  District  2917  1910635  
4             97          84          0  District  2917  1910635

## District Summary

In [537]:
# Calculate the total number of unique schools
school_count = len(school_data_complete["school_name"].unique())
school_count

15

In [538]:
# Calculate the total number of students
student_count = len(school_data_complete["student_name"])

#Formatting the values
student_total = f"{student_count:,}"
print(student_total)

#Printing original answer before formatting
student_count

39,170


39170

In [539]:
# Calculate the total budget
total_budget = school_data_complete["budget"].sum()

#Formatting the values
total_budget_formatted = f"${total_budget:,}"
print(total_budget_formatted)

#Printing original answer before formatting
total_budget

$82,932,329,558


82932329558

In [540]:
# Calculate the average (mean) math score
average_math_score = school_data_complete["math_score"].mean()

#Formatting the values
average_score_math_formatted = f"{average_math_score:.2f}"
print(average_score_math_formatted)

#Printing original answer before formatting
average_math_score

78.99


78.98537145774827

In [541]:
# Calculate the average (mean) reading score
average_reading_score = school_data_complete["reading_score"].mean()

#Formatting the values
average_reading_score_formatted = f"{average_reading_score:.2f}"
print(average_reading_score_formatted)

#Printing original answer before formatting
average_reading_score

81.88


81.87784018381414

In [542]:
# Use the following to calculate the percentage of students who passed math (math scores greather than or equal to 70)
passing_math_count = school_data_complete[(school_data_complete["math_score"] >= 70)].count()["student_name"]
passing_math_percentage = passing_math_count / np.int64(student_count) * 100

#Formatting the values
passing_math_percentage_formatted = f"{passing_math_percentage:.2f}%"
print(passing_math_percentage_formatted)

#Printing original answer before formatting
passing_math_percentage

81.88%


74.9808526933878

In [543]:
# Calculate the percentage of students who passed reading (hint: look at how the math percentage was calculated)  
passing_reading_count = school_data_complete[(school_data_complete["reading_score"] >= 70)].count()["student_name"]
passing_reading_percentage = passing_reading_count / np.int64(student_count) * 100

#Formatting the values
passing_reading_percentage_formatted = f"{average_reading_score:.2f}%"
print(passing_reading_percentage_formatted)

#Printing original answer before formatting
passing_reading_percentage

81.88%


85.80546336482001

In [544]:
# Use the following to calculate the percentage of students that passed math and reading
passing_math_reading_count = school_data_complete[
    (school_data_complete["math_score"] >= 70) & (school_data_complete["reading_score"] >= 70)
].count()["student_name"]
overall_passing_rate = passing_math_reading_count /  float(student_count) * 100

#Formatting the values
overall_passing_rate_formatted = f"{overall_passing_rate:.2f}%"
print(overall_passing_rate_formatted)

#Printing original answer before formatting
overall_passing_rate

65.17%


65.17232575950983

In [545]:
# Create a high-level snapshot of the district's key metrics in a DataFrame
district_summary = pd.DataFrame({"Total Schools" : school_count, "Total Number of Students" : student_total,
                                 "Total Budget" : total_budget_formatted, 
                                 "Average Score (Maths)" : average_score_math_formatted,
                                 "Average Score (Reading)" : average_reading_score_formatted,
                                 "% Passing (Maths)" : passing_math_percentage_formatted,
                                 "% Passing(Reading)" : passing_reading_percentage_formatted,
                                 "Overall Passing %" : overall_passing_rate_formatted}, 
                                index=[0])

# Formatting
# district_summary["Total Students"] = district_summary["Total Students"].map("{:,}".format)
# district_summary["Total Budget"] = district_summary["Total Budget"].map("${:,.2f}".format)

# Display the DataFrame
district_summary

Total Schools Total Number of Students     Total Budget  \
0             15                   39,170  $82,932,329,558   

  Average Score (Maths) Average Score (Reading) % Passing (Maths)  \
0                 78.99                   81.88            81.88%   

  % Passing(Reading) Overall Passing %  
0             81.88%            65.17%

## School Summary

In [546]:
#For reference
school_data_complete.head()

Student ID       student_name gender grade        school_name  \
0           0       Paul Bradley      M   9th  Huang High School   
1           1       Victor Smith      M  12th  Huang High School   
2           2    Kevin Rodriguez      M  12th  Huang High School   
3           3  Dr. Richard Scott      M  12th  Huang High School   
4           4         Bonnie Ray      F   9th  Huang High School   

   reading_score  math_score  School ID      type  size   budget  
0             66          79          0  District  2917  1910635  
1             94          61          0  District  2917  1910635  
2             90          60          0  District  2917  1910635  
3             67          58          0  District  2917  1910635  
4             97          84          0  District  2917  1910635

In [547]:
#Setting the School name as the index for the School Summary table
grouped_schools = school_data_complete.groupby(["school_name"])

In [548]:
# Use the code provided to select all of the school types
school_types = grouped_schools["type"].first()

In [549]:
# Calculate the total student count per school
per_school_counts = school_data_complete["school_name"].value_counts()

In [550]:
# Calculating the total school budget per school
per_school_budget = grouped_schools["budget"].first()

In [551]:
#Calculating the per capita spending per school
per_school_capita = per_school_budget/per_school_counts

In [552]:
# Calculate the average test scores (Math) per school
per_school_math = grouped_schools["math_score"].mean()

In [553]:
# Calculate the average test scores (Reading) per school
per_school_reading = grouped_schools["reading_score"].mean()

In [554]:
# Calculate the number of students per school with math scores of 70 or higher
students_passing_math = school_data_complete[(school_data_complete["math_score"] >= 70)]
school_students_passing_math = students_passing_math.groupby("school_name")["student_name"].count()

In [555]:
# Calculate the number of students per school with reading scores of 70 or higher
students_passing_reading = school_data_complete[(school_data_complete["reading_score"] >= 70)]
school_students_passing_reading = students_passing_math.groupby("school_name")["student_name"].count()

In [556]:
# Use the provided code to calculate the number of students per school that passed both math and reading with scores of 70 or higher
students_passing_math_and_reading = school_data_complete[
    (school_data_complete["reading_score"] >= 70) & (school_data_complete["math_score"] >= 70)
]
school_students_passing_math_and_reading = students_passing_math_and_reading.groupby(["school_name"]).size()

In [557]:
# Use the provided code to calculate the passing rates
per_school_passing_math = school_students_passing_math / per_school_counts * 100
per_school_passing_reading = school_students_passing_reading / per_school_counts * 100
overall_passing_rate = school_students_passing_math_and_reading / per_school_counts * 100

In [558]:
# Create a DataFrame called `per_school_summary` with columns for the calculations above.
per_school_summary = pd.DataFrame({"School Type" : school_types, "Total Number of Students" : per_school_counts, 
                                   "Total Budget" : per_school_budget, 
                                   "Per Capita Spending" : per_school_capita, 
                                   "Average Score (Maths)" : per_school_math,
                                   "Average Score (Reading)" : per_school_reading, 
                                   "% Passing (Math)" : per_school_passing_math,
                                   "% Passing (Reading)" : per_school_passing_reading,
                                   "Overall Passing %" : overall_passing_rate})

# Formatting
per_school_summary["Total Number of Students"] = per_school_summary["Total Number of Students"].map("{:,}".format)
per_school_summary["Total Budget"] = per_school_summary["Total Budget"].map("${:,.2f}".format)
per_school_summary["Per Capita Spending"] = per_school_summary["Per Capita Spending"].map("${:,.2f}".format)
per_school_summary["Average Score (Maths)"] = per_school_summary["Average Score (Maths)"].map("{:,.2f}".format)
per_school_summary["Average Score (Reading)"] = per_school_summary["Average Score (Reading)"].map("{:,.2f}".format)
per_school_summary["% Passing (Math)"] = per_school_summary["% Passing (Math)"].map("{:.2f}%".format)
per_school_summary["% Passing (Reading)"] = per_school_summary["% Passing (Reading)"].map("{:.2f}%".format)
per_school_summary["Overall Passing %"] = per_school_summary["Overall Passing %"].map("{:.2f}%".format)

# Display the DataFrame
per_school_summary

School Type Total Number of Students   Total Budget  \
Bailey High School       District                    4,976  $3,124,928.00   
Cabrera High School       Charter                    1,858  $1,081,356.00   
Figueroa High School     District                    2,949  $1,884,411.00   
Ford High School         District                    2,739  $1,763,916.00   
Griffin High School       Charter                    1,468    $917,500.00   
Hernandez High School    District                    4,635  $3,022,020.00   
Holden High School        Charter                      427    $248,087.00   
Huang High School        District                    2,917  $1,910,635.00   
Johnson High School      District                    4,761  $3,094,650.00   
Pena High School          Charter                      962    $585,858.00   
Rodriguez High School    District                    3,999  $2,547,363.00   
Shelton High School       Charter                    1,761  $1,056,600.00   
Thomas High School        Charter                    1,635  $1,043,130.00   
Wilson High School        Charter                    2,283  $1,319,574.00   
Wright High School        Charter                    1,800  $1,049,400.00   

                      Per Capita Spending Average Score (Maths)  \
Bailey High School                $628.00                 77.05   
Cabrera High School               $582.00                 83.06   
Figueroa High School              $639.00                 76.71   
Ford High School                  $644.00                 77.10   
Griffin High School               $625.00                 83.35   
Hernandez High School             $652.00                 77.29   
Holden High School                $581.00                 83.80   
Huang High School                 $655.00                 76.63   
Johnson High School               $650.00                 77.07   
Pena High School                  $609.00                 83.84   
Rodriguez High School             $637.00                 76.84   
Shelton High School               $600.00                 83.36   
Thomas High School                $638.00                 83.42   
Wilson High School                $578.00                 83.27   
Wright High School                $583.00                 83.68   

                      Average Score (Reading) % Passing (Math)  \
Bailey High School                      81.03           66.68%   
Cabrera High School                     83.98           94.13%   
Figueroa High School                    81.16           65.99%   
Ford High School                        80.75           68.31%   
Griffin High School                     83.82           93.39%   
Hernandez High School                   80.93           66.75%   
Holden High School                      83.81           92.51%   
Huang High School                       81.18           65.68%   
Johnson High School                     80.97           66.06%   
Pena High School                        84.04           94.59%   
Rodriguez High School                   80.74           66.37%   
Shelton High School                     83.73           93.87%   
Thomas High School                      83.85           93.27%   
Wilson High School                      83.99           93.87%   
Wright High School                      83.95           93.33%   

                      % Passing (Reading) Overall Passing %  
Bailey High School                 66.68%            54.64%  
Cabrera High School                94.13%            91.33%  
Figueroa High School               65.99%            53.20%  
Ford High School                   68.31%            54.29%  
Griffin High School                93.39%            90.60%  
Hernandez High School              66.75%            53.53%  
Holden High School                 92.51%            89.23%  
Huang High School                  65.68%            53.51%  
Johnson High School                66.06%            53.54%  
Pena High School                   94.5

## Highest-Performing Schools (by % Overall Passing)

In [559]:
# Sort the schools by `% Overall Passing` in descending order and display the top 5 rows.
top_schools = per_school_summary.sort_values("Overall Passing %", ascending = False)
top_schools.head(5)

School Type Total Number of Students   Total Budget  \
Cabrera High School     Charter                    1,858  $1,081,356.00   
Thomas High School      Charter                    1,635  $1,043,130.00   
Griffin High School     Charter                    1,468    $917,500.00   
Wilson High School      Charter                    2,283  $1,319,574.00   
Pena High School        Charter                      962    $585,858.00   

                    Per Capita Spending Average Score (Maths)  \
Cabrera High School             $582.00                 83.06   
Thomas High School              $638.00                 83.42   
Griffin High School             $625.00                 83.35   
Wilson High School              $578.00                 83.27   
Pena High School                $609.00                 83.84   

                    Average Score (Reading) % Passing (Math)  \
Cabrera High School                   83.98           94.13%   
Thomas High School                    83.85           93.27%   
Griffin High School                   83.82           93.39%   
Wilson High School                    83.99           93.87%   
Pena High School                      84.04           94.59%   

                    % Passing (Reading) Overall Passing %  
Cabrera High School              94.13%            91.33%  
Thomas High School               93.27%            90.95%  
Griffin High School              93.39%            90.60%  
Wilson High School               93.87%            90.58%  
Pena High School                 94.59%            90.54%

## Bottom Performing Schools (By % Overall Passing)

In [560]:
# Sort the schools by `% Overall Passing` in ascending order and display the top 5 rows.
bottom_schools = per_school_summary.sort_values("Overall Passing %")
bottom_schools.head(5)

School Type Total Number of Students   Total Budget  \
Rodriguez High School    District                    3,999  $2,547,363.00   
Figueroa High School     District                    2,949  $1,884,411.00   
Huang High School        District                    2,917  $1,910,635.00   
Hernandez High School    District                    4,635  $3,022,020.00   
Johnson High School      District                    4,761  $3,094,650.00   

                      Per Capita Spending Average Score (Maths)  \
Rodriguez High School             $637.00                 76.84   
Figueroa High School              $639.00                 76.71   
Huang High School                 $655.00                 76.63   
Hernandez High School             $652.00                 77.29   
Johnson High School               $650.00                 77.07   

                      Average Score (Reading) % Passing (Math)  \
Rodriguez High School                   80.74           66.37%   
Figueroa High School                    81.16           65.99%   
Huang High School                       81.18           65.68%   
Hernandez High School                   80.93           66.75%   
Johnson High School                     80.97           66.06%   

                      % Passing (Reading) Overall Passing %  
Rodriguez High School              66.37%            52.99%  
Figueroa High School               65.99%            53.20%  
Huang High School                  65.68%            53.51%  
Hernandez High School              66.75%            53.53%  
Johnson High School                66.06%            53.54%

## Math Scores by Grade

In [561]:
school_data_complete.head()

Student ID       student_name gender grade        school_name  \
0           0       Paul Bradley      M   9th  Huang High School   
1           1       Victor Smith      M  12th  Huang High School   
2           2    Kevin Rodriguez      M  12th  Huang High School   
3           3  Dr. Richard Scott      M  12th  Huang High School   
4           4         Bonnie Ray      F   9th  Huang High School   

   reading_score  math_score  School ID      type  size   budget  
0             66          79          0  District  2917  1910635  
1             94          61          0  District  2917  1910635  
2             90          60          0  District  2917  1910635  
3             67          58          0  District  2917  1910635  
4             97          84          0  District  2917  1910635

In [562]:
# Use the code provided to separate the data by grade
ninth_graders = school_data_complete[(school_data_complete["grade"] == "9th")]
tenth_graders = school_data_complete[(school_data_complete["grade"] == "10th")]
eleventh_graders = school_data_complete[(school_data_complete["grade"] == "11th")]
twelfth_graders = school_data_complete[(school_data_complete["grade"] == "12th")]

# Group by `school_name` and take the mean of the `math_score` column for each.
ninth_grade_math_scores = ninth_graders.groupby("school_name")["math_score"].mean().map("{:,.4f}".format)
tenth_grader_math_scores = tenth_graders.groupby("school_name")["math_score"].mean().map("{:,.4f}".format)
eleventh_grader_math_scores = eleventh_graders.groupby("school_name")["math_score"].mean().map("{:,.4f}".format)
twelfth_grader_math_scores = twelfth_graders.groupby("school_name")["math_score"].mean().map("{:,.4f}".format)

# Combine each of the scores above into single DataFrame called `math_scores_by_grade`
math_scores_by_grade = pd.DataFrame({"9th Grade": ninth_grade_math_scores, "10th Grade": tenth_grader_math_scores,
                                    "11th Grade": eleventh_grader_math_scores,"12th Grade": twelfth_grader_math_scores})

# Minor data wrangling
math_scores_by_grade.index.name = "Math Scores per Grade"

# Display the DataFrame
math_scores_by_grade

9th Grade 10th Grade 11th Grade 12th Grade
Math Scores per Grade                                           
Bailey High School      77.0837    76.9968    77.5156    76.4922
Cabrera High School     83.0947    83.1545    82.7656    83.2775
Figueroa High School    76.4030    76.5400    76.8843    77.1514
Ford High School        77.3613    77.6723    76.9181    76.1800
Griffin High School     82.0440    84.2291    83.8421    83.3562
Hernandez High School   77.4385    77.3374    77.1360    77.1866
Holden High School      83.7874    83.4298    85.0000    82.8554
Huang High School       77.0273    75.9087    76.4466    77.2256
Johnson High School     77.1879    76.6911    77.4917    76.8632
Pena High School        83.6255    83.3720    84.3281    84.1215
Rodriguez High School   76.8600    76.6125    76.3956    77.6907
Shelton High School     83.4208    82.9174    83.3835    83.7790
Thomas High School      83.5900    83.0879    83.4988    83.4970
Wilson High School      83.0856    83.7244    83.1953    83.0358
Wright High School      83.2647    84.0103    83.8368    83.6450

## Reading Score by Grade 

In [563]:
# Use the code provided to separate the data by grade
ninth_graders = school_data_complete[(school_data_complete["grade"] == "9th")]
tenth_graders = school_data_complete[(school_data_complete["grade"] == "10th")]
eleventh_graders = school_data_complete[(school_data_complete["grade"] == "11th")]
twelfth_graders = school_data_complete[(school_data_complete["grade"] == "12th")]

# Group by `school_name` and take the mean of the the `reading_score` column for each.
ninth_grade_reading_scores = ninth_graders.groupby("school_name")["reading_score"].mean().map("{:,.4f}".format)
tenth_grader_reading_scores = tenth_graders.groupby("school_name")["reading_score"].mean().map("{:,.4f}".format)
eleventh_grader_reading_scores = eleventh_graders.groupby("school_name")["reading_score"].mean().map("{:,.4f}".format)
twelfth_grader_reading_scores = twelfth_graders.groupby("school_name")["reading_score"].mean().map("{:,.4f}".format)

# Combine each of the scores above into single DataFrame called `reading_scores_by_grade`
reading_scores_by_grade = pd.DataFrame({"9th": ninth_grade_reading_scores, "10th": tenth_grader_reading_scores,
                                        "11th": eleventh_grader_reading_scores, "12th": twelfth_grader_reading_scores})

# Minor data wrangling
reading_scores_by_grade = reading_scores_by_grade[["9th", "10th", "11th", "12th"]]
reading_scores_by_grade.index.name = "Reading Scores per Grade"

# Display the DataFrame
reading_scores_by_grade

9th     10th     11th     12th
Reading Scores per Grade                                    
Bailey High School        81.3032  80.9072  80.9456  80.9125
Cabrera High School       83.6761  84.2532  83.7884  84.2880
Figueroa High School      81.1986  81.4089  80.6403  81.3849
Ford High School          80.6327  81.2627  80.4036  80.6623
Griffin High School       83.3692  83.7069  84.2881  84.0137
Hernandez High School     80.8669  80.6601  81.3961  80.8571
Holden High School        83.6772  83.3246  83.8155  84.6988
Huang High School         81.2903  81.5124  81.4175  80.3060
Johnson High School       81.2607  80.7734  80.6160  81.2276
Pena High School          83.8073  83.6120  84.3359  84.5912
Rodriguez High School     80.9931  80.6298  80.8648  80.3764
Shelton High School       84.1226  83.4420  84.3738  82.7817
Thomas High School        83.7289  84.2542  83.5855  83.8314
Wilson High School        83.9398  84.0215  83.7646  84.3177
Wright High School        83.8333  83.8128  84.1563  84.0732

## Scores by School Spending

In [564]:
# Establish the bins 
spending_bins = [0, 585, 630, 645, 680]
labels = ["<$585", "$585-630", "$630-645", "$645-680"]

In [565]:
# Create a copy of the school summary since it has the "Per Student Budget" 
school_spending_df = per_school_summary.copy()

In [566]:
# Use `pd.cut` to categorize spending based on the bins.
school_spending_df["Spending Ranges (Per Student)"] = pd.cut(per_school_budget/per_school_counts,
                                                             spending_bins, labels=labels)
school_spending_df

School Type Total Number of Students   Total Budget  \
Bailey High School       District                    4,976  $3,124,928.00   
Cabrera High School       Charter                    1,858  $1,081,356.00   
Figueroa High School     District                    2,949  $1,884,411.00   
Ford High School         District                    2,739  $1,763,916.00   
Griffin High School       Charter                    1,468    $917,500.00   
Hernandez High School    District                    4,635  $3,022,020.00   
Holden High School        Charter                      427    $248,087.00   
Huang High School        District                    2,917  $1,910,635.00   
Johnson High School      District                    4,761  $3,094,650.00   
Pena High School          Charter                      962    $585,858.00   
Rodriguez High School    District                    3,999  $2,547,363.00   
Shelton High School       Charter                    1,761  $1,056,600.00   
Thomas High School        Charter                    1,635  $1,043,130.00   
Wilson High School        Charter                    2,283  $1,319,574.00   
Wright High School        Charter                    1,800  $1,049,400.00   

                      Per Capita Spending Average Score (Maths)  \
Bailey High School                $628.00                 77.05   
Cabrera High School               $582.00                 83.06   
Figueroa High School              $639.00                 76.71   
Ford High School                  $644.00                 77.10   
Griffin High School               $625.00                 83.35   
Hernandez High School             $652.00                 77.29   
Holden High School                $581.00                 83.80   
Huang High School                 $655.00                 76.63   
Johnson High School               $650.00                 77.07   
Pena High School                  $609.00                 83.84   
Rodriguez High School             $637.00                 76.84   
Shelton High School               $600.00                 83.36   
Thomas High School                $638.00                 83.42   
Wilson High School                $578.00                 83.27   
Wright High School                $583.00                 83.68   

                      Average Score (Reading) % Passing (Math)  \
Bailey High School                      81.03           66.68%   
Cabrera High School                     83.98           94.13%   
Figueroa High School                    81.16           65.99%   
Ford High School                        80.75           68.31%   
Griffin High School                     83.82           93.39%   
Hernandez High School                   80.93           66.75%   
Holden High School                      83.81           92.51%   
Huang High School                       81.18           65.68%   
Johnson High School                     80.97           66.06%   
Pena High School                        84.04           94.59%   
Rodriguez High School                   80.74           66.37%   
Shelton High School                     83.73           93.87%   
Thomas High School                      83.85           93.27%   
Wilson High School                      83.99           93.87%   
Wright High School                      83.95           93.33%   

                      % Passing (Reading) Overall Passing %  \
Bailey High School                 66.68%            54.64%   
Cabrera High School                94.13%            91.33%   
Figueroa High School               65.99%            53.20%   
Ford High School                   68.31%            54.29%   
Griffin High School                93.39%            90.60%   
Hernandez High School              66.75%            53.53%   
Holden High School                 92.51%            89.23%   
Huang High School                  65.68%            53.51%   
Johnson High School                66.06%            53.54%   
Pena High School             

In [567]:
school_spending_df["Average Score (Maths)"] = school_spending_df["Average Score (Maths)"].apply(lambda x: x.replace(',', '')).astype('float')
school_spending_df["Average Score (Reading)"] = school_spending_df["Average Score (Reading)"].apply(lambda x: x.replace(',', '')).astype('float')
school_spending_df["% Passing (Math)"] = school_spending_df["% Passing (Math)"].apply(lambda x: x.replace('%', '').replace(',', '')).astype('float')
school_spending_df["% Passing (Reading)"] = school_spending_df["% Passing (Reading)"].apply(lambda x: x.replace('%', '').replace(',', '')).astype('float')
school_spending_df["Overall Passing %"] = school_spending_df["Overall Passing %"].apply(lambda x: x.replace('%', '').replace(',', '')).astype('float')

In [568]:
#  Calculate averages for the desired columns. 
grouped_data_df = school_spending_df.groupby(["Spending Ranges (Per Student)"])

spending_math_scores = grouped_data_df["Average Score (Maths)"].mean()

spending_reading_scores = grouped_data_df["Average Score (Reading)"].mean()

spending_passing_math = grouped_data_df["% Passing (Math)"].mean()

spending_passing_reading = grouped_data_df["% Passing (Reading)"].mean()

overall_passing_spending = grouped_data_df["Overall Passing %"].mean()

In [569]:
# Assemble into DataFrame
spending_summary = pd.DataFrame({"Average Score (Maths)":spending_math_scores,
                                 "Average Score (Reading)":spending_reading_scores,
                                 "% Passing (Math)":spending_passing_math,
                                 "% Passing (Reading)":spending_passing_reading, 
                                 "Overall Passing %":overall_passing_spending})

spending_summary.index.name = "Scores by School Spending"

# Display results
spending_summary

Average Score (Maths)  Average Score (Reading)  \
Scores by School Spending                                                   
<$585                                  83.452500                83.932500   
$585-630                               81.900000                83.155000   
$630-645                               78.517500                81.625000   
$645-680                               76.996667                81.026667   

                           % Passing (Math)  % Passing (Reading)  \
Scores by School Spending                                          
<$585                             93.460000            93.460000   
$585-630                          87.132500            87.132500   
$630-645                          73.485000            73.485000   
$645-680                          66.163333            66.163333   

                           Overall Passing %  
Scores by School Spending                     
<$585                              90.367500  
$585-630                           81.417500  
$630-645                           62.857500  
$645-680                           53.526667

## Scores by School Size

In [570]:
# Establish the bins.
size_bins = [0, 1000, 2000, 5000]
labels = ["Small (<1000)", "Medium (1000-2000)", "Large (2000-5000)"]

In [571]:
# Categorize the spending based on the bins
# Use `pd.cut` on the "Total Students" column of the `per_school_summary` DataFrame.

per_school_summary["School Size"] = pd.cut(school_data_complete["school_name"].value_counts(),
                                           size_bins, labels=labels)
                                           
per_school_summary

School Type Total Number of Students   Total Budget  \
Bailey High School       District                    4,976  $3,124,928.00   
Cabrera High School       Charter                    1,858  $1,081,356.00   
Figueroa High School     District                    2,949  $1,884,411.00   
Ford High School         District                    2,739  $1,763,916.00   
Griffin High School       Charter                    1,468    $917,500.00   
Hernandez High School    District                    4,635  $3,022,020.00   
Holden High School        Charter                      427    $248,087.00   
Huang High School        District                    2,917  $1,910,635.00   
Johnson High School      District                    4,761  $3,094,650.00   
Pena High School          Charter                      962    $585,858.00   
Rodriguez High School    District                    3,999  $2,547,363.00   
Shelton High School       Charter                    1,761  $1,056,600.00   
Thomas High School        Charter                    1,635  $1,043,130.00   
Wilson High School        Charter                    2,283  $1,319,574.00   
Wright High School        Charter                    1,800  $1,049,400.00   

                      Per Capita Spending Average Score (Maths)  \
Bailey High School                $628.00                 77.05   
Cabrera High School               $582.00                 83.06   
Figueroa High School              $639.00                 76.71   
Ford High School                  $644.00                 77.10   
Griffin High School               $625.00                 83.35   
Hernandez High School             $652.00                 77.29   
Holden High School                $581.00                 83.80   
Huang High School                 $655.00                 76.63   
Johnson High School               $650.00                 77.07   
Pena High School                  $609.00                 83.84   
Rodriguez High School             $637.00                 76.84   
Shelton High School               $600.00                 83.36   
Thomas High School                $638.00                 83.42   
Wilson High School                $578.00                 83.27   
Wright High School                $583.00                 83.68   

                      Average Score (Reading) % Passing (Math)  \
Bailey High School                      81.03           66.68%   
Cabrera High School                     83.98           94.13%   
Figueroa High School                    81.16           65.99%   
Ford High School                        80.75           68.31%   
Griffin High School                     83.82           93.39%   
Hernandez High School                   80.93           66.75%   
Holden High School                      83.81           92.51%   
Huang High School                       81.18           65.68%   
Johnson High School                     80.97           66.06%   
Pena High School                        84.04           94.59%   
Rodriguez High School                   80.74           66.37%   
Shelton High School                     83.73           93.87%   
Thomas High School                      83.85           93.27%   
Wilson High School                      83.99           93.87%   
Wright High School                      83.95           93.33%   

                      % Passing (Reading) Overall Passing %  \
Bailey High School                 66.68%            54.64%   
Cabrera High School                94.13%            91.33%   
Figueroa High School               65.99%            53.20%   
Ford High School                   68.31%            54.29%   
Griffin High School                93.39%            90.60%   
Hernandez High School              66.75%            53.53%   
Holden High School                 92.51%            89.23%   
Huang High School                  65.68%            53.51%   
Johnson High School                66.06%            53.54%   
Pena High School             

In [572]:
per_school_summary["Average Score (Maths)"] = per_school_summary["Average Score (Maths)"].apply(lambda x: x.replace(',', '')).astype('float')
per_school_summary["Average Score (Reading)"] = per_school_summary["Average Score (Reading)"].apply(lambda x: x.replace(',', '')).astype('float')
per_school_summary["% Passing (Math)"] = per_school_summary["% Passing (Math)"].apply(lambda x: x.replace('%', '').replace(',', '')).astype('float')
per_school_summary["% Passing (Reading)"] = per_school_summary["% Passing (Reading)"].apply(lambda x: x.replace('%', '').replace(',', '')).astype('float')
per_school_summary["Overall Passing %"] = per_school_summary["Overall Passing %"].apply(lambda x: x.replace('%', '').replace(',', '')).astype('float')

In [573]:
# Calculate averages for the desired columns. 
size_math_scores = per_school_summary.groupby(["School Size"])["Average Score (Maths)"].mean()
size_reading_scores = per_school_summary.groupby(["School Size"])["Average Score (Reading)"].mean()
size_passing_math = per_school_summary.groupby(["School Size"])["% Passing (Math)"].mean()
size_passing_reading = per_school_summary.groupby(["School Size"])["% Passing (Reading)"].mean()
size_overall_passing = per_school_summary.groupby(["School Size"])["Overall Passing %"].mean()

In [574]:
# Create a DataFrame called `size_summary` that breaks down school performance based on school size (small, medium, or large).
# Use the scores above to create a new DataFrame called `size_summary`
size_summary = pd.DataFrame({"Average Score (Maths)":size_math_scores,
                             "Average Score (Reading)":size_reading_scores,
                             "% Passing (Math)":size_passing_math,
                             "% Passing (Reading)":size_passing_reading,
                             "Overall Passing %":size_overall_passing})

size_summary.index.name = "Scores by School Size"

# Display results
size_summary

Average Score (Maths)  Average Score (Reading)  \
Scores by School Size                                                   
Small (<1000)                         83.820                 83.92500   
Medium (1000-2000)                    83.374                 83.86600   
Large (2000-5000)                     77.745                 81.34375   

                       % Passing (Math)  % Passing (Reading)  \
Scores by School Size                                          
Small (<1000)                  93.55000             93.55000   
Medium (1000-2000)             93.59800             93.59800   
Large (2000-5000)              69.96375             69.96375   

                       Overall Passing %  
Scores by School Size                     
Small (<1000)                     89.885  
Medium (1000-2000)                90.620  
Large (2000-5000)                 58.285

## Scores by School Type

In [575]:
# Group the per_school_summary DataFrame by "School Type" and average the results.
average_math_score_by_type = per_school_summary.groupby(["School Type"])["Average Score (Maths)"].mean()
average_reading_score_by_type = per_school_summary.groupby(["School Type"])["Average Score (Reading)"].mean()
average_percent_passing_math_by_type = per_school_summary.groupby(["School Type"])["% Passing (Math)"].mean()
average_percent_passing_reading_by_type = per_school_summary.groupby(["School Type"])["% Passing (Reading)"].mean()
average_percent_overall_passing_by_type = per_school_summary.groupby(["School Type"])["Overall Passing %"].mean()

In [576]:
# Assemble the new data by type into a DataFrame called `type_summary`
type_summary = pd.DataFrame({"Average Score (Maths)":average_math_score_by_type,
                             "Average Score (Reading)":average_reading_score_by_type,
                             "% Passing (Math)":average_percent_passing_math_by_type,
                             "% Passing (Reading)":average_percent_passing_reading_by_type,
                             "Overall Passing %":average_percent_overall_passing_by_type})

type_summary.index.name = "Scores by School Type"

# Display results
type_summary

Average Score (Maths)  Average Score (Reading)  \
Scores by School Type                                                   
Charter                            83.472500                83.896250   
District                           76.955714                80.965714   

                       % Passing (Math)  % Passing (Reading)  \
Scores by School Type                                          
Charter                       93.620000            93.620000   
District                      66.548571            66.548571   

                       Overall Passing %  
Scores by School Type                     
Charter                        90.431250  
District                       53.671429

### Analysis Summary
From the above analysis, we can summarize that the highest overall passing % goes to those schools that:-
+ have less Per Capita Spending
    + eg: Those schools that allocates $645 to $680 for each student has an overall passing of approximately 53.53%
>Those schools that allocate <$585 for each student has an overall passing of approximately 90.37%
+ are medium sized
    + eg: Those schools that are medium sized have a passing percent of 90.62% while the other groups are lower
    > Large = 58.285% and Small = 89.885%
+ are Chartered
    + Chartered schools have a pass percent of 90.43%
    + District schools only have a percentage of 53.67%